crates/abi/abi/HEVM.sol

```diff
@@ -84,6 +86,9 @@ record()
 accesses(address)(bytes32[], bytes32[])
 skip(bool)
 
+recordStateDiff()
+getStateDiff()(AccountAccess[])
```

Comment:
Depends on where the record was initiated. If you start recording just prior to `A.foo()`, then there wouldn't be any `AccountAccess.A` since A wasn't accessed. The first AccountAccess would be for B, with A as its _accessor_. So the returned array would be as you say, without the `AccountAccess.A`.

Yeah, I'm starting to agree with you on a new AccessKind needed for this return control flow. We do need to maintain the "chronological" order so `StorageAccess C, slot 3` relates with `StorageAccess B, slot 5`. But it won't be clear to users that the latter storage access isn't from a call.

OTOH, this kinda defeats the purpose of having a `storageAccesses` field for `AccountAccess` if we need a new kind to handle this case. I'm not exactly sure the right approach should be. But I lean towards completely decoupling storage accesses from account accesses, while maintaining ordering. So you have a the following state diff array:
- AccountAccess.B
- StorageAccess.B, slot 0
- AccountAccess.C
- StorageAccess.C, slot 3
- StorageAccess.B, slot 5

The main issue with this approach is that the user cannot tell the context `StorageAccess.B, slot 5` is for. This _should_ be evident from the `StorageAccess.account` field, but perhaps in cases like delegatecall it's a bit blurry.
This API depends on whether users need to know exactly which context an account access occurred.

Reply:
> Depends on where the record was initiated. If you start recording just prior to `A.foo()`, then there wouldn't be any `AccountAccess.A` since A wasn't accessed. The first AccountAccess would be for B, with A as its _accessor_. So the returned array would be as you say, without the `AccountAccess.A`.

Interesting, I'd expect this to have resulted in `AccountAccess.A` entry:
```solidity
contract MyTest is Test {
  function testAccesses() external {
    vm.recordStateDiff();
    a.foo(); // `foo` now calls `b`
    AccountAccess[] memory accesses = vm.getStateDiff();
  }
}
```
And I think it should, because users will want the storage accesses that took place in the `a.foo()` call also

---

Hmm, yea this is tricky. A "return control flow" seems ok but feels hacky. And I don't think Solidity is flexible enough to allow the cheat to return a flat array of `(AccountAccess | StorageAccess)[]` so the UX there may be clunky because I think you'd need it return a struct of:
```solidity
struct AccountOrStorageAccess {
  AccountAccess accountAccess; // empty if it was a storage access
  StorageAccess storageAccess; // empty if it was an account access
  bool isAccountAccess;
  bool isStorageAccess;
}
```

Reply:
Let me clarify how recording occurs. In your code snippet, the `a.foo()` CALL will be recorded. What I was trying to explain is that state accesses in the same context a record begins are not recorded. That is the `sstore` in the following example won't be recorded:
```solidity
contract MyTest is Test {
  function testAccesses() external {
    vm.recordStateDiff();
    assembly { sstore(0x0, 0x0) }
    a.foo(); // `foo` now calls `b`
    AccountAccess[] memory accesses = vm.getStateDiff();
  }
}
```
The `a.foo()` call generates an `AccountAccess` record for only the `b` contract, but not `a` since that's the only account being accessed there. I guess, if there is a value transfer then it should generate two records for the credit and debit changes. Though this will have to be designated with a new type of account access or even a new Access type struct.

Reply:
In your example, is the `a.foo()` an actual CALL, i.e. calling an external A.foo function for example? If so, then I think we're on the same page. If it's just a jump, then that kinda breaks the CALL kind concept we have going.

Reply:
Oh sorry it was a CALL, I see how that is unclear

Reply:
Just noting for completeness here that we discussed offline and agreed on adding a synthetic `Resume` access type to handle the "control flow returned to caller" access kind